ngraph/core/src/op/einsum.cpp

```diff
@@ -0,0 +1,277 @@
+                                    std::string& output_subscript)
+{
+    // remove extra white-spaces from the equation
+    m_equation.erase(std::remove_if(m_equation.begin(), m_equation.end(), ::isspace),
+                     m_equation.end());
+
+    // split equation to input subscripts and an output subscript
+    auto pos_output_delimeter = m_equation.find("->");
+    auto input_subscripts_str = m_equation.substr(0, pos_output_delimeter);
+
+    // split the input subscripts into a vector of input subscripts
+    input_subscripts.clear();
+    std::istringstream input;
+    input.str(input_subscripts_str);
+    for (std::string input_subscript; std::getline(input, input_subscript, ',');)
+    {
+        input_subscripts.push_back(input_subscript);
+    }
+
+    // check that each input subscript contains only alphabetic letter or ellipsis
+    const std::regex subscript_regex("[A-Za-z]*|[A-Za-z]*\\.\\.\\.[A-Za-z]*");
+    for (auto const& input_subscript : input_subscripts)
+    {
+        if (std::regex_match(input_subscript, subscript_regex) == false)
+        {
+            return false;
+        }
+    }
+
+    // figure out if ellipsis is met in input subscripts
+    const std::regex ellipsis_regex("\\.\\.\\.");
+    bool is_ellipsis_met = false;
+    std::smatch ellipsis_match;
+    for (auto const& input_subscript : input_subscripts)
+    {
+        if (std::regex_search(input_subscript, ellipsis_match, ellipsis_regex) == true)
+        {
+            is_ellipsis_met = true;
+            break;
+        }
+    }
+
+    if (pos_output_delimeter == std::string::npos)
+    {
+        // recover output subscript
+        output_subscript = "";
+        for (auto const& input_subscript : input_subscripts)
+        {
+            for (auto const& label : input_subscript)
+            {
+                if (std::isalpha(label) && output_subscript.find(label) == std::string::npos)
+                {
+                    output_subscript += label;
+                }
+            }
+        }
+        std::sort(output_subscript.begin(), output_subscript.end());
+        if (is_ellipsis_met)
+        {
+            output_subscript = "..." + output_subscript;
+        }
+    }
+    else
+    {
+        output_subscript = m_equation.substr(pos_output_delimeter + 2);
+
+        // check that the output subscript has the correct format
+        if (std::regex_match(output_subscript, subscript_regex) == false)
+        {
+            return false;
+        }
+
+        // if the ellipsis is met in input subscripts, one ellipsis must be in the output subscript
+        std::regex_search(output_subscript, ellipsis_match, ellipsis_regex);
+        if (is_ellipsis_met && ellipsis_match.size() != 1)
+        {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+void op::v7::Einsum::extract_labels(std::string const& subscript, std::vector<std::string>& labels)
+{
+    labels.clear();
```

Comment:
may be you should do labels as method result?

Reply:
I think it is better to return simple types data but it is case of `vector` type. Of course, we can use `move` but here I avoid it.